Nemolite2d omp offload

vkatkinson opened the pull request:
Added functionality for OpenMP GPU offloading.

rupertford commented:
@vkatkinson, could you advise on whether you are you are still working on this PR or whether it is ready for review.

vkatkinson commented:
Hi @rupertford, it's ready for review.

rupertford commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/time_step_omp_gpu.cpp:
```diff
@@ -0,0 +1,156 @@
+        double g
+  	){
+
+#pragma omp target enter data map (to: area_v[0:width*width], ssha_u[0:width*width], sshn_t[0:width*width], sshn_u[0:width*width], sshn_v[0:width*width], ssha_t[0:width*width], hu[0:width*width], hv[0:width*width], un[0:width*width], vn[0:width*width], area_t[0:width*width], ht[0:width*width], tmask[0:width*width], gphiu[0:width*width], dx_u[0:width*width], dx_v[0:width*width], dx_t[0:width*width ...
```
Is width*width correct? Doesn't that assume a square domain?

vkatkinson replied:
@rupertford I've now changed this to width*height and updated time_step_mod.f90 accordingly.

rupertford commented on compiler_setup/nvidia.sh:
```diff
@@ -0,0 +1,51 @@
+# Build settings for the Nvidia compiler
```
I think the two pgi scripts should be merged or removed at some point.

vkatkinson replied:
I've now removed both the pgi.sh and pgi_acc.sh compiler set up files.

rupertford reviewed (changes requested):
Good job. I can get the examples to compile and run using gnu - although the GPU offload will do nothing on my machine. I've not tested the llvm and nvidia as I'm having problems logging on to glados so I will take these on trust (and because I've seen you run them). Comments are inline.

vkatkinson commented on compiler_setup/nvidia.sh:
```diff
@@ -0,0 +1,51 @@
+
+# Flags to use when compiling with OpenMP support
+OMPFLAGS="-mp"
+# Flags to use when compiling with OpenMP support
```
I've updated these comments.

vkatkinson commented on compiler_setup/nvidia.sh:
```diff
@@ -0,0 +1,51 @@
+# Registers are shared by threads in an SMP. The more registers a kernel
+# uses, the fewer threads it can support. This parameter can be tuned and
+# shoul be a multiple of 8.
+# -Mcuda is required to build CUDA Fortran
```
I've removed these comments.

vkatkinson commented on compiler_setup/nvidia.sh:
```diff
@@ -0,0 +1,51 @@
+# cc = compute capability
+# Registers are shared by threads in an SMP. The more registers a kernel
+# uses, the fewer threads it can support. This parameter can be tuned and
+# shoul be a multiple of 8.
```
I've now removed these comments.

vkatkinson commented on compiler_setup/nvidia.sh:
```diff
@@ -0,0 +1,51 @@
+CXX=nvc++
+
+# C and C++ flags
+CFLAGS="-O3 -Minfo=all -g"
```
Both flags aren't necessary but provide useful debugging information so I've kept them for now.

rupertford replied:
Fine, but please add a comment explaining this. Is this file also an experimental target (as it is not mentioned as being so)? If not, and someone was going to run this themselves to get performance results, should they include these flags or not? Do they affect performance or not?

rupertford replied:
You also have a separate commented-out set of debugging options. So what is the difference. Are there two debug levels you use? Could you have some words to explain to someone who looked at the file and didn't know what options to use.

vkatkinson replied:
I've removed these from the C and F90 flags and added a debugging options comment below detailing what could be used if needed. These debugging flags would affect performance so you're right that they shouldn't be included here.

vkatkinson commented on compiler_setup/llvm.sh:
```diff
@@ -0,0 +1,41 @@
+LDFLAGS="-lomp -lomptarget"
+# Location of various CUDA maths libraries
+#LDFLAGS+=" -L${CUDA_MATH_DIR}/lib64"
+LDFLAGS+=" -L/apps/packages/compilers/llvm/13.0.0rc3/lib"
```
I've removed the hardcoded path.

vkatkinson commented on compiler_setup/llvm.sh:
```diff
@@ -0,0 +1,41 @@
+CXX=clang++
+
+# C and C++ flags
+CFLAGS="-O3 -march=native -g"
```
The -g flag provides debugging information so I've kept it for now. It may be appropriate to delete it at a later stage.

rupertford replied:
Could you please explain in the file, not just to me here. I would expect a comment at the top of the file, or above the relevant lines explaining that -g is kept for debugging as this is an experimental implementation.

vkatkinson replied:
I've added a comment by the C and F90 flags explaining why -g is included.

vkatkinson commented on compiler_setup/llvm.sh:
```diff
@@ -0,0 +1,41 @@
+# C and C++ flags
+CFLAGS="-O3 -march=native -g"
+# Fortran compiler flags
+F90FLAGS="-O3 -march=native -g"
```
As above

rupertford replied:
As above :-)

vkatkinson commented on compiler_setup/llvm.sh:
```diff
@@ -0,0 +1,41 @@
+# Flags to use when compiling with OpenMP support
+OMPFLAGS="-fopenmp"
+# Flags to use when compiling with OpenMP support
+#OMPTARGETFLAGS="–fopenmp-targets=nvptx64-nvidia-cuda"
```
This file is experimental so other flags may be useful. I've kept the commented version in this case and added a comment at the top of the file to explain why different variations of flags are provided.

rupertford replied:
OK, but could you please be more explicit in the explanation. Perhaps saying something like ... "In some cases, where they have been useful, potential alternative options are provided as comments."

vkatkinson replied:
I've now added a further comment at the top of the script.

vkatkinson commented on compiler_setup/llvm.sh:
```diff
@@ -0,0 +1,41 @@
+# Linker flags
+LDFLAGS="-lomp -lomptarget"
+# Location of various CUDA maths libraries
+#LDFLAGS+=" -L${CUDA_MATH_DIR}/lib64"
```
I've now uncommented this version.

vkatkinson commented on compiler_setup/nvidia.sh:
```diff
@@ -0,0 +1,51 @@
+# C and C++ flags
+CFLAGS="-O3 -Minfo=all -g"
+# Fortran compiler flags
+F90FLAGS="-O3 -Minfo=all"
```
As above

rupertford replied:
As above :-)

vkatkinson replied:
As above

vkatkinson commented on compiler_setup/nvidia.sh:
```diff
@@ -0,0 +1,51 @@
+F90FLAGS="-O3 -Minfo=all"
+# Debugging options
+#F90FLAGS"+=" -fcheck=all -fbacktrace -ffpe-trap=invalid -g -O0"
+# -Mcuda is for CUDA Fortran
```
I've now removed the comments which are not directly related to the flags.

vkatkinson commented:
Hi @rupertford, thanks for the feedback. I've addressed all of your comments.

rupertford reviewed (changes requested):
Thanks @vkatkinson. Nearly there. I'd just like to see some more explanation in the options files which would let someone new to the code(s) and compilers to know what options to use and why. Remaining comments are inline.

vkatkinson commented:
Thanks @rupertford. I've added in some brief information about the debugging flags and provided further context about the LLVM set up script. Now that I've removed the debugging flags from the C and F90 flags in `nvidia.sh`, do you think it's worth keeping the debugging options comment below or assume that these can be looked up in the compiler documentation if needed?

rupertford commented:
> ... Now that I've removed the debugging flags from the C and F90 flags in `nvidia.sh`, do you think it's worth keeping the debugging options comment below or assume that these can be looked up in the compiler documentation if needed?

Yes I think keeping potentially useful options for other users (and ourselves when we come back to it in the future) is definitely a good thing. You would have gotten away with a list and a statement saying to add these for debugging but I have no complaints about the additional detail :-)

rupertford reviewed (approved):
All issues have been addressed. Approving.
